File: engine/engine-model/src/main/scala/server/maps/AbstractMapleMapObject.scala
package server.maps

import java.awt.Point

abstract class AbstractMapleMapObject() extends MapleMapObject {
  private var _position: Option[Point] = Option.apply(new Point())

  private var _objectId: Int = 0

  override def objectId: Int = _objectId

  override def objectId_$eq(value: Int): Unit = _objectId = value

  def position: Point = {
    new Point(_position.get)
  }

  override def position_$eq(position: Point): Unit = {
    if (_position.isDefined) {
      _position.get.x = position.x
      _position.get.y = position.y
    }
  }

  override def nullifyPosition(): Unit = _position = Option.empty
}

File: engine/engine-model/src/main/java/server/maps/AbstractMapleMapObject.java
package server.maps;

import java.awt.Point;

public abstract class AbstractMapleMapObject implements MapleMapObject {
   private Point position;

   private Integer objectId;

   @Override
   public int objectId() {
      return objectId;
   }

   @Override
   public void setObjectId(int id) {
      objectId = id;
   }

   @Override
   public Point position() {
      return new Point(position);
   }

   @Override
   public void setPosition(Point position) {
      if (this.position != null) {
         this.position.move(position.x, position.y);
      }
   }

   @Override
   public void nullifyPosition() {
      position = null;
   }
}
